contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,268 @@
+
+    /// ========== INTERNAL FUNCTIONS ==========
+
+    /**
+     * @notice stake sushi rewards
+     */
+    function _stakeSushi() internal {
+        uint256 balance = IERC20(sushi).balanceOf(address(this));
+        if (balance > 0) {
+            IERC20(sushi).approve(xSushi, balance);
+            ISushiBar(xSushi).enter(balance); // stake sushi
+        }
+    }
+
+    /**
+     * @notice unStake sushi rewards
+     */
+    function _unStakeSushi() internal {
+        uint256 balance = IERC20(xSushi).balanceOf(address(this));
+        if (balance > 0) {
+            ISushiBar(xSushi).leave(balance); // unstake $xSUSHI
+        }
+    }
+
+    /**
+     *  @notice pending $SUSHI rewards
+     *  @return uint
+     */
+    function _pendingSushi(uint256 pid) internal view returns (uint256) {
+        return IMasterChef(masterChef).pendingSushi(pid, address(this));
+    }
+
+    /**
+     * @notice Find out pool Id from Onsen based on the LP token
+     * according to Onsen documentation there shouldn't be more than pool by LPToken, so we are going to use the first ocurrence of the LPToken in the Onsen pools
+     * @return [bool, uint256], first value shows if value was found, second value the id for the pool
+     */
+    function _findPoolByLP(address LPToken) internal view returns (bool, uint256) {
+        /// Check if the id is already stored in the local variable
+        PoolData memory pd = _lpToOnsenId[LPToken];
+
+        /// If the pool is active return the id else search for the pool id
+        if (pd.poolActive) {
+            return (pd.poolActive, pd.id);
+        } else {
+            /// Call the poolLength function from sushi masterchef v2 and compare against the LP token passed as parameter
+            uint256 poolsLength = IMasterChef(masterChef).poolLength();
+
+            for (uint256 i; i < poolsLength; i++) {
+                /// If found, return the the pool Id from Onsen for the current LPToken.
+                if (LPToken == (address)(IMasterChef(masterChef).lpToken(i))) {
+                    pd = _lpToOnsenId[LPToken];
+                    pd.poolActive = true;
+                    pd.id = i;
+                    return (true, i);
+                }
+            }
+
+            /// If the LPToken was not found return 0 and indicate that the token was not found
+            return (false, 0);
+        }
+    }
+
+    function _deallocateAll() internal {
+        uint256[] memory amounts = new uint256[](_tokens.length);
+
+        // interactions
+        for (uint256 i; i < _tokens.length; i++) {
+            (bool onsenLPFound, uint256 onsenPoolId) = _findPoolByLP(address(_tokens[i]));
+            if (onsenLPFound) {
+                /// Retrieve current balance for pool and address
+                UserInfo memory currentUserInfo = IMasterChef(masterChef).userInfo(onsenPoolId, address(this));
+                amounts[i] = currentUserInfo.amount;
+            }
+        }
+
+        deallocate(amounts);
```

Comment:
This seems unnecessary layer of function calling to me. We can just call withdraw and harvest in our first for loop instead of delegating to this function and looping again.

Reply:
Yes, you are right, the function was updated to call the withdrawAndHarvest in this same loop, avoiding looping it again.